In [3]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf
import matplotlib.pyplot as plt
from scipy import stats

In [7]:
df_casa = pd.read_excel('./Ejemplo_Casa.xls')
display(df_casa.head())

PRECIO  LOTE  CUARTOS  BANOS  PISOS  ENTRADA  REC  SOTANO  CALEF  AIRE  \
0   42000  5850        3      1      2        1    0       1      0     0   
1   38500  4000        2      1      1        1    0       0      0     0   
2   49500  3060        3      1      1        1    0       0      0     0   
3   60500  6650        3      1      2        1    1       0      0     0   
4   61000  6360        2      1      1        1    0       0      0     0   

   GARAGE  NBHD  
0       1     0  
1       0     0  
2       0     0  
3       0     0  
4       0     0

In [8]:
y=df_casa['PRECIO']
X=sm.add_constant(df_casa.drop(columns='PRECIO'))

reg_casa=sm.OLS(y,X).fit()
display(reg_casa.summary())

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 PRECIO   R-squared:                       0.673
Model:                            OLS   Adj. R-squared:                  0.666
Method:                 Least Squares   F-statistic:                     99.97
Date:                Mon, 28 Jul 2025   Prob (F-statistic):          6.18e-122
Time:                        10:31:58   Log-Likelihood:                -6034.1
No. Observations:                 546   AIC:                         1.209e+04
Df Residuals:                     534   BIC:                         1.214e+04
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -4038.3504   3409.471     -1.184      0.237   -1.07e+04    2659.271
LOTE           3.5463      0.350     10.124      0.000       2.858       4.234
CUARTOS     1832.0035   1047.000      1.750      0.081    -224.741    3888.748
BANOS       1.434e+04   1489.921      9.622      0.000    1.14e+04    1.73e+04
PISOS       6556.9457    925.290      7.086      0.000    4739.291    8374.600
ENTRADA     6687.7789   2045.246      3.270      0.001    2670.065    1.07e+04
REC         4511.2838   1899.958      2.374      0.018     778.976    8243.592
SOTANO      5452.3855   1588.024      3.433      0.001    2332.845    8571.926
CALEF       1.283e+04   3217.597      3.988      0.000    6510.706    1.92e+04
AIRE        1.263e+04   1555.021      8.124      0.000    9578.182    1.57e+04
GARAGE      4244.8290    840.544      5.050      0.000    2593.650    5896.008
NBHD        9369.5132   1669.091      5.614      0.000    6090.724    1.26e+04
==============================================================================
Omnibus:                       93.454   Durbin-Watson:                   1.604
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              247.620
Skew:                           0.853   Prob(JB):                     1.70e-54
Kurtosis:                       5.824   Cond. No.                     3.07e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.07e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### Hacemos las modificaciones al df para quitar las observaciones, variables dummy, etc

In [17]:
df_casa_filtrado=df_casa[df_casa['BANOS']!=4].copy()
y1=df_casa_filtrado['PRECIO']
X1=sm.add_constant(df_casa_filtrado.drop(columns='PRECIO'))
display(X1)
reg_casa_filtrado=sm.OLS(y1,X1).fit()
display(reg_casa_filtrado.summary())

const  LOTE  CUARTOS  BANOS  PISOS  ENTRADA  REC  SOTANO  CALEF  AIRE  \
0      1.0  5850        3      1      2        1    0       1      0     0   
1      1.0  4000        2      1      1        1    0       0      0     0   
2      1.0  3060        3      1      1        1    0       0      0     0   
3      1.0  6650        3      1      2        1    1       0      0     0   
4      1.0  6360        2      1      1        1    0       0      0     0   
..     ...   ...      ...    ...    ...      ...  ...     ...    ...   ...   
541    1.0  4800        3      2      4        1    1       0      0     1   
542    1.0  6000        3      2      4        1    0       0      0     1   
543    1.0  6000        3      2      4        1    1       0      0     1   
544    1.0  6000        3      2      2        1    1       0      0     1   
545    1.0  6000        3      1      2        1    0       0      0     1   

     GARAGE  NBHD  
0         1     0  
1         0     0  
2         0     0  
3         0     0  
4         0     0  
..      ...   ...  
541       0     0  
542       0     0  
543       1     0  
544       1     0  
545       1     0  

[545 rows x 12 columns]

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 PRECIO   R-squared:                       0.665
Model:                            OLS   Adj. R-squared:                  0.658
Method:                 Least Squares   F-statistic:                     96.14
Date:                Mon, 28 Jul 2025   Prob (F-statistic):          7.67e-119
Time:                        10:35:36   Log-Likelihood:                -6022.2
No. Observations:                 545   AIC:                         1.207e+04
Df Residuals:                     533   BIC:                         1.212e+04
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -3509.7393   3419.390     -1.026      0.305   -1.02e+04    3207.395
LOTE           3.5341      0.350     10.102      0.000       2.847       4.221
CUARTOS     1909.3244   1046.422      1.825      0.069    -146.293    3964.941
BANOS       1.383e+04   1519.347      9.102      0.000    1.08e+04    1.68e+04
PISOS       6492.6879    924.670      7.022      0.000    4676.244    8309.132
ENTRADA     6722.7877   2042.146      3.292      0.001    2711.146    1.07e+04
REC         4623.2425   1898.206      2.436      0.015     894.360    8352.125
SOTANO      5506.6270   1585.876      3.472      0.001    2391.293    8621.961
CALEF         1.3e+04   3214.103      4.043      0.000    6681.465    1.93e+04
AIRE        1.262e+04   1552.591      8.130      0.000    9572.753    1.57e+04
GARAGE      4139.8173    841.671      4.919      0.000    2486.418    5793.216
NBHD        9469.9610   1667.598      5.679      0.000    6194.090    1.27e+04
==============================================================================
Omnibus:                       97.367   Durbin-Watson:                   1.595
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              263.699
Skew:                           0.881   Prob(JB):                     5.48e-58
Kurtosis:                       5.917   Cond. No.                     3.08e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.08e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [10]:
print(reg_casa_filtrado.params[0])

-3509.739310860323


/var/folders/jr/t_k402pj3bg57t7p56rgmknr0000gn/T/ipykernel_2271/3109356172.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(reg_casa_filtrado.params[0])


### a) Interpretacion del coeficiente 4139.8 de la variable GARAGE
Cuando el espacio sube en 1 unidad el precio de las casas se incrementa en  4139.8

### b) Utilidad del estadístico F

El estadístico F sirve para ver si en su conjunto el modelo es significativo. La hipótesis nula es que todas las regresoras son =0  y la alternativa es que todas son distintas de cero. En este ejemplo el estadístico F es singnificativo lo que indica que el modelo globalmente esta bien, pero como hay coeficimentes no significativos, es posible que exista multicolinealidad.

Calculo: 

F=(RRSS-URSS)/q /(URSS/(1-k))

In [11]:
print(f"F-statistic: {reg_casa_filtrado.fvalue}")
print(f"p-value del test F: {reg_casa_filtrado.f_pvalue}")

F-statistic: 96.13889762772537
p-value del test F: 7.67254444765226e-119


### El p-value es bajo, lo que implica que es muy significativo, por o tanto el modelo esta bien especificado en su conjunto y como hay esitmadores no signficativos, es posible que exista multicolinealidad.

### C)
El hecho de que las varianzas seas grandes es un signo de multicolinealidad. 
La multicolinealidad perfecta hace que no se puedan obtener los estimadores $\beta$, sin embargo lo que sucede normalmente es un grado de multicolinealidad, por lo tanto se obtienen los $\beta$, siguen siendo MELI, pero son malos estimadores, eso se ve en la varianza o std.
Otro indicativo es el número de condicion que es alto.
Otro indicador mas el que exista un t poco significativo y un R^2 ajustado alto. 
Aca en general, si bien el número de cuartos tiene un nivel de significancial al 10% y no al 5% o mas como las otras variables, no parece haber un problema de multicolinealidad puesto que el estadístico F es muy significativo. Quitar cuatros generaría un problema de especificación, pero si valdria el análisis entre cuatros y banos, es posible que exista una relación entre ambos.


### D) tenemos que predecir

In [33]:
df_prediccion=pd.DataFrame({'LOTE': [5100],'CUARTOS': [3],'BANOS': [2],'PISOS':[2],'ENTRADA':[1],'REC':[1],'SOTANO':[0],'CALEF':[1],'AIRE':[0],'GARAGE':[0],'NBHD':[1]})
X_prediccion = sm.add_constant(df_prediccion,has_constant='add')

display(X_prediccion)
display(X)
prediccion=reg_casa_filtrado.predict(X_prediccion)
display(prediccion)

const  LOTE  CUARTOS  BANOS  PISOS  ENTRADA  REC  SOTANO  CALEF  AIRE  \
0    1.0  5100        3      2      2        1    1       0      1     0   

   GARAGE  NBHD  
0       0     1

const  LOTE  CUARTOS  BANOS  PISOS  ENTRADA  REC  SOTANO  CALEF  AIRE  \
0      1.0  5850        3      1      2        1    0       1      0     0   
1      1.0  4000        2      1      1        1    0       0      0     0   
2      1.0  3060        3      1      1        1    0       0      0     0   
3      1.0  6650        3      1      2        1    1       0      0     0   
4      1.0  6360        2      1      1        1    0       0      0     0   
..     ...   ...      ...    ...    ...      ...  ...     ...    ...   ...   
541    1.0  4800        3      2      4        1    1       0      0     1   
542    1.0  6000        3      2      4        1    0       0      0     1   
543    1.0  6000        3      2      4        1    1       0      0     1   
544    1.0  6000        3      2      2        1    1       0      0     1   
545    1.0  6000        3      1      2        1    0       0      0     1   

     GARAGE  NBHD  
0         1     0  
1         0     0  
2         0     0  
3         0     0  
4         0     0  
..      ...   ...  
541       0     0  
542       0     0  
543       1     0  
544       1     0  
545       1     0  

[546 rows x 12 columns]

0    94697.639528
dtype: float64